src/main/utility/synchronization/semaphore.rs

```diff
@@ -0,0 +1,205 @@
+use std::cell::UnsafeCell;
+use std::sync::Arc;
+
+/// A libc semaphore that provides synchronization between threads. Any memory barrier/ordering
```

Comment:
I guess there is not a rust equivalent in the std lib? Maybe worth explaining why we need our own (optional).